kvprober: add leaseholder information for ranges in kvprober logs

ShashwatDash23 opened the pull request:
Previously, leaseholder information for ranges probed by kvprober was not being logged.

This commit introduces logging of leaseholder details extracted from kvprober traces. Currently, getting leaseholder information is a time consuming process. This aims to improve the debugging process by providing readily accessible leaseholder data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126507)
<!-- Reviewable:end -->

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=126507) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=126507) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=126507) it.</sub>

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -306,11 +308,42 @@ func (p *Prober) Metrics() Metrics {
+
+		// Extract all occurrences of range IDs and leaseholder node IDs from the log message
+		ranges := rangeRegex.FindAllStringSubmatch(informationLog, -1)
+		leaseholders := leaseRegex.FindAllStringSubmatch(informationLog, -1)
```
Not 100% confident, but I think in the case of the prober, we expect exactly one leaseholder to be involved, since only a single KV pair in KV is probed. When you run your prototype, do you see multiple leaseholders in the logs or one?

ShashwatDash23 replied:
Yes, the line I have used to get information seems to sometimes get information for the ranges just before and after the range being probed as shown in the comment below.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -306,11 +308,42 @@ func (p *Prober) Metrics() Metrics {
+// and logs this information only if the leaseholder for a range has been updated.
+func (p *Prober) LogLeaseholderInfo(ctx context.Context, recording tracingpb.Recording) {
+	// Find the log message containing information about updated ranges
+	informationLog, isValid := recording.FindLogMessage("received updated range")
```
Can you paste the textual representation of the trace into this comment thread, so I can review this code with that in mind?

ShashwatDash23 replied:
`event:kv/kvclient/kvcoord/dist_sender.go:2872 [n3,kvprober,txn=‹17d32b7a›] received updated range info: [‹desc: r63:/Table/6{1-2} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212389000,1, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r62:/Table/6{0-1} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212290000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r64:/Table/6{2-3} [(n1,s1):1, (n2,s2):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.234352000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING›]`

This is the line I used to get the lease information

ShashwatDash23 replied:
```
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8       0.000ms      0.000ms    === operation:read probe loop _verbose:‹1› node:‹3› kvprober:
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.000ms      0.000ms    [/cockroach.roachpb.Internal/Batch: {count: 6, duration 2ms}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.000ms      0.000ms    [dist sender send: {count: 3, duration 2ms}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.000ms      0.000ms    [txn coordinator send: {count: 1, duration 767µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.021ms      0.021ms        === operation:dist sender send _verbose:‹1› node:‹3› kvprober:
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.021ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 1ms}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.071ms      0.050ms        event:kv/kvclient/kvcoord/range_iter.go:188 [n3,kvprober] querying next range at /Meta2/‹""›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.105ms      0.034ms        event:kv/kvclient/kvcoord/range_iter.go:225 [n3,kvprober] key: /Meta2/‹""›, desc: r1:‹/{Min-System/NodeLiveness}› [(n1,s1):1, (n2,s2):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.125ms      0.021ms        event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober] r1: sending batch ‹1 Scan› to (n1,s1):1
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.137ms      0.011ms        event:rpc/nodedialer/nodedialer.go:147 [n3,kvprober] sending request to ‹localhost:26257›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.144ms      0.007ms        event:kv/kvclient/kvcoord/transport.go:213 [n3,kvprober] sending batch request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.146ms      0.002ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹3› kvprober: span.kind:‹client›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.321ms      0.175ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹1› span.kind:‹server› request:‹Scan [/Meta2/"",/Meta2/"\xff\xff\x00"), [max_span_request_keys: 100], [target_bytes: 0]›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.332ms      0.011ms            event:server/node.go:1407 [n1] node received request: ‹1 Scan›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.353ms      0.021ms            event:kv/kvserver/store_send.go:154 [n1,s1] executing Scan [/Meta2/‹""›,/Meta2/‹"\xff\xff\x00"›), [max_span_request_keys: 100], [target_bytes: 0]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.360ms      0.007ms            event:kv/kvserver/replica_send.go:176 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read-only path
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.367ms      0.007ms            event:kv/kvserver/concurrency/concurrency_manager.go:214 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] sequencing request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.371ms      0.004ms            event:kv/kvserver/concurrency/concurrency_manager.go:295 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] acquiring latches
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.377ms      0.006ms            event:kv/kvserver/concurrency/concurrency_manager.go:339 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] scanning lock table for conflicting locks
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.406ms      0.029ms            event:kv/kvserver/replica_read.go:313 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] can drop latches early for batch (Scan [/Meta2/‹""›,/Meta2/‹"\xff\xff\x00"›), [max_span_request_keys: 100], [target_bytes: 0]); scanning lock table first to detect conflicts
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.463ms      0.057ms            event:kv/kvserver/replica_read.go:125 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] lock table scan complete without conflicts; dropping latches early
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.484ms      0.021ms            event:kv/kvserver/replica_read.go:426 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] executing read-only batch
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.580ms      0.096ms            event:kv/kvserver/replica_evaluate.go:516 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] evaluated Scan command ‹header:<kvnemesis_seq:<> key:"\003" end_key:"\003\377\377\000" > ›, txn=<nil> : resp=‹num_keys:68 num_bytes:4031 ›, err=<nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.605ms      0.025ms            event:n1 scan stats: stepped 68 times (130 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 130, key-bytes 1.6 KiB, value-bytes 5.0 KiB, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.616ms      0.011ms            event:kv/kvserver/replica_read.go:238 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read completed
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.625ms      0.009ms            event:server/node.go:1501 [n1] node sending response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.584ms     -0.041ms            structured:‹{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"1","numInternalSeeks":"1","numInterfaceSteps":"68","numInternalSteps":"130","keyBytes":"1614","valueBytes":"5104","pointCount":"130","numScans":"1","nodeId":1}›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.862ms      0.719ms        event:kv/kvclient/kvcoord/transport.go:215 [n3,kvprober] received batch response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.888ms      0.026ms        === operation:dist sender send _verbose:‹1› node:‹3› kvprober:
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.888ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 528µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.898ms      0.010ms        event:kv/kvclient/kvcoord/range_iter.go:188 [n3,kvprober] querying next range at /Meta2/‹""›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.914ms      0.016ms        event:kv/kvclient/kvcoord/range_iter.go:225 [n3,kvprober] key: /Meta2/‹""›, desc: r1:‹/{Min-System/NodeLiveness}› [(n1,s1):1, (n2,s2):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.924ms      0.010ms        event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober] r1: sending batch ‹1 Scan› to (n1,s1):1
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.929ms      0.005ms        event:rpc/nodedialer/nodedialer.go:147 [n3,kvprober] sending request to ‹localhost:26257›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.934ms      0.005ms        event:kv/kvclient/kvcoord/transport.go:213 [n3,kvprober] sending batch request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.939ms      0.005ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹3› kvprober: span.kind:‹client›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.024ms      0.086ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹1› span.kind:‹server› request:‹Scan [/Meta2/"",/Meta2/"\xff\xff\x00"), [max_span_request_keys: 32], [target_bytes: 0]›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.030ms      0.006ms            event:server/node.go:1407 [n1] node received request: ‹1 Scan›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.042ms      0.012ms            event:kv/kvserver/store_send.go:154 [n1,s1] executing Scan [/Meta2/‹""›,/Meta2/‹"\xff\xff\x00"›), [max_span_request_keys: 32], [target_bytes: 0]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.048ms      0.006ms            event:kv/kvserver/replica_send.go:176 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read-only path
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.055ms      0.007ms            event:kv/kvserver/concurrency/concurrency_manager.go:217 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] optimistically sequencing request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.071ms      0.016ms            event:kv/kvserver/concurrency/concurrency_manager.go:289 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] optimistically acquiring latches
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.083ms      0.012ms            event:kv/kvserver/concurrency/concurrency_manager.go:335 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] optimistically scanning lock table for conflicting locks
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.091ms      0.008ms            event:kv/kvserver/replica_read.go:426 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] executing read-only batch
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.145ms      0.054ms            event:kv/kvserver/replica_evaluate.go:516 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] evaluated Scan command ‹header:<kvnemesis_seq:<> key:"\003" end_key:"\003\377\377\000" > ›, txn=<nil> : resp=‹resume_span:<key:"\003\247" end_key:"\003\377\377\000" > resume_reason:RESUME_KEY_LIMIT num_keys:32 num_bytes:1935 ›, err=<nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.158ms      0.013ms            event:n1 scan stats: stepped 64 times (175 internal); seeked 2 times (2 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 177, key-bytes 3.9 KiB, value-bytes 7.3 KiB, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.175ms      0.017ms            event:kv/kvserver/replica_read.go:238 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read completed
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.183ms      0.008ms            event:server/node.go:1501 [n1] node sending response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.147ms     -0.036ms            structured:‹{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"2","numInternalSeeks":"2","numInterfaceSteps":"64","numInternalSteps":"175","keyBytes":"4013","valueBytes":"7424","pointCount":"177","numScans":"1","nodeId":1}›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.326ms      0.393ms        event:kv/kvclient/kvcoord/transport.go:215 [n3,kvprober] received batch response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.378ms      0.051ms        === operation:txn coordinator send _verbose:‹1› node:‹3› kvprober: txnID:‹17d32b7a-cb42-4e2c-8738-798ec2f7e1fb›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.378ms      0.000ms        [dist sender send: {count: 1, duration 740µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.378ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 630µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.382ms      0.004ms            === operation:dist sender send _verbose:‹1› node:‹3› kvprober: txn:‹17d32b7a›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.382ms      0.000ms            [/cockroach.roachpb.Internal/Batch: {count: 2, duration 630µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.394ms      0.012ms            event:kv/kvclient/kvcoord/range_iter.go:188 [n3,kvprober,txn=‹17d32b7a›] querying next range at /Table/61
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.411ms      0.018ms            event:kv/kvclient/kvcoord/range_iter.go:225 [n3,kvprober,txn=‹17d32b7a›] key: /Table/61, desc: r63:‹/Table/6{1-2}› [(n1,s1):1, next=2, gen=0]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.438ms      0.027ms            event:kv/kvclient/kvcoord/dist_sender.go:2579 [n3,kvprober,txn=‹17d32b7a›] routing to nearest replica; leaseholder not known
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.447ms      0.009ms            event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober,txn=‹17d32b7a›] r63: sending batch ‹1 Get› to (n1,s1):1
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.454ms      0.006ms            event:rpc/nodedialer/nodedialer.go:147 [n3,kvprober,txn=‹17d32b7a›] sending request to ‹localhost:26257›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.460ms      0.006ms            event:kv/kvclient/kvcoord/transport.go:213 [n3,kvprober,txn=‹17d32b7a›] sending batch request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.461ms      0.001ms                === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹3› kvprober: txn:‹17d32b7a› span.kind:‹client›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.590ms      0.129ms                === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹1› span.kind:‹server› request:‹Get [/Local/Range/Table/61/RangeProbe], [txn: 17d32b7a], [can-forward-ts]›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.600ms      0.010ms                event:server/node.go:1407 [n1] node received request: ‹1 Get›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.620ms      0.020ms                event:kv/kvserver/store_send.go:154 [n1,s1] executing Get [/Local/Range/Table/61/‹RangeProbe›], [txn: 17d32b7a], [can-forward-ts]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.632ms      0.012ms                event:kv/kvserver/replica_send.go:176 [n1,s1,r63/1:‹/Table/6{1-2}›] read-only path
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.640ms      0.008ms                event:kv/kvserver/concurrency/concurrency_manager.go:214 [n1,s1,r63/1:‹/Table/6{1-2}›] sequencing request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.645ms      0.005ms                event:kv/kvserver/concurrency/concurrency_manager.go:295 [n1,s1,r63/1:‹/Table/6{1-2}›] acquiring latches
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.651ms      0.006ms                event:kv/kvserver/concurrency/concurrency_manager.go:339 [n1,s1,r63/1:‹/Table/6{1-2}›] scanning lock table for conflicting locks
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.666ms      0.015ms                event:kv/kvserver/replica_read.go:313 [n1,s1,r63/1:‹/Table/6{1-2}›] can drop latches early for batch (Get [/Local/Range/Table/61/‹RangeProbe›], [txn: 17d32b7a], [can-forward-ts]); scanning lock table first to detect conflicts
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.677ms      0.011ms                event:kv/kvserver/replica_read.go:125 [n1,s1,r63/1:‹/Table/6{1-2}›] lock table scan complete without conflicts; dropping latches early
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.684ms      0.007ms                event:kv/kvserver/replica_read.go:426 [n1,s1,r63/1:‹/Table/6{1-2}›] executing read-only batch
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.720ms      0.036ms                event:kv/kvserver/replica_evaluate.go:516 [n1,s1,r63/1:‹/Table/6{1-2}›] evaluated Get command ‹header:<kvnemesis_seq:<> key:"\001k\022\305\000\001prbe" > ›, txn="unnamed" meta={id=17d32b7a key=/Min iso=Serializable pri=0.01077494 epo=0 ts=1719946576.032320000,0 min=1719946576.032320000,0 seq=0} lock=false stat=PENDING rts=1719946576.032320000,0 wto=false gul=1719946576.532320000,0 : resp=‹header:<> ›, err=<nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.730ms      0.010ms                event:n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 1 gets, 0 scans, 0 reverse scans
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.739ms      0.009ms                event:kv/kvserver/replica_read.go:238 [n1,s1,r63/1:‹/Table/6{1-2}›] read completed
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.751ms      0.012ms                event:kv/kvserver/replica_send.go:331 [n1,s1,r63/1:‹/Table/6{1-2}›] client had stale range info; returning an update
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.772ms      0.021ms                event:server/node.go:1501 [n1] node sending response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.721ms     -0.051ms                structured:‹{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"1","numInternalSeeks":"1","numGets":"1","nodeId":1}›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.923ms      0.463ms            event:kv/kvclient/kvcoord/transport.go:215 [n3,kvprober,txn=‹17d32b7a›] received batch response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.025ms      0.102ms            event:kv/kvclient/kvcoord/dist_sender.go:2872 [n3,kvprober,txn=‹17d32b7a›] received updated range info: [‹desc: r63:/Table/6{1-2} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212389000,1, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r62:/Table/6{0-1} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212290000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r64:/Table/6{2-3} [(n1,s1):1, (n2,s2):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.234352000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING›]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.053ms      0.028ms            event:kv/kvclient/rangecache/range_cache.go:1022 [n3,kvprober,txn=‹17d32b7a›] evict cached descriptor: desc=‹desc:r63:/Table/6{1-2} [(n1,s1):1, next=2, gen=0], lease:<empty>›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.062ms      0.009ms            event:kv/kvclient/rangecache/range_cache.go:526 [n3,kvprober,txn=‹17d32b7a›] evicting cached range descriptor with 3 replacements
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.082ms      0.019ms            event:kv/kvclient/rangecache/range_cache.go:1196 [n3,kvprober,txn=‹17d32b7a›] clearing entries overlapping r63:‹/Table/6{1-2}› [(n1,s1):1, (n3,s3):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.099ms      0.018ms            event:kv/kvclient/rangecache/range_cache.go:1196 [n3,kvprober,txn=‹17d32b7a›] clearing entries overlapping r62:‹/Table/6{0-1}› [(n1,s1):1, (n3,s3):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.119ms      0.020ms            event:kv/kvclient/rangecache/range_cache.go:1196 [n3,kvprober,txn=‹17d32b7a›] clearing entries overlapping r64:‹/Table/6{2-3}› [(n1,s1):1, (n2,s2):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.143ms      0.766ms        event:kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:661 [n3,kvprober,txn=‹17d32b7a›] recording span to refresh: ‹/Local/Range/Table/61/RangeProbe›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.164ms      2.164ms    event:kv/txn.go:1062 [n3,kvprober] kv.Txn did AutoCommit. err: <nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.181ms      0.017ms    event:kv/kvprober/kvprober.go:470 [n3,kvprober] kv.Get(/Local/Range/Table/61/‹RangeProbe›), r=63 returned success in 797.417µs

```

This is the entire trace

joshimhoff replied:
I think you should pull the leaseholder out of this part of the trace:

```
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.447ms      0.009ms            event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober,txn=‹17d32b7a›] r63: sending batch ‹1 Get› to (n1,s1):1
```

We want the final occurrence of above in the trace. The earlier occurrences are related to the planning happening in `planner.go`, instead of the probing of the target range happening in `kvprober.go`'s `readProbeImpl`, etc. methods. We are interested in the leaseholder of the target range specifically.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -306,11 +308,42 @@ func (p *Prober) Metrics() Metrics {
+			newCtx = logtags.AddTag(newCtx, "leaseholder", leaseholders[i][1])
+
+			// Log the range and leaseholder information using the new context
+			log.Health.Infof(newCtx, "range %s has leaseholder %s", ranges[i][1], leaseholders[i][1])
```
I think instead of logging via a separate log line, we should include the leaseholder in existing logs lines such as what is below. So perhaps this function should return the leaseholder. Can be called from `readProbeImpl`, etc.

```
	log.Health.Infof(ctx, "kv.Get(%s), r=%v returned success in %v", step.Key, step.RangeID, d)
```

This will make it possible to do Datadog queries, where we break down the error rate by leaseholder. Similar to: https://us5.datadoghq.com/logs?query=service%3Acockroachdb%20%40friendly_id%3Acrl-prod-n89%20status%3Aerror%20%22unable%20to%20connect%20%28is%20the%20peer%20up%20and%20reachable%3F%29%22%20&agg_m=count&agg_m_source=base&agg_q=%40tags.n&agg_q_source=base&agg_t=count&analyticsOptions=%5B%22bars%22%2C%22dog_classic%22%2Cnull%2Cnull%5D&cols=host%2Cservice%2C%40tags.n%2C%40tags.rnode&fromUser=true&messageDisplay=inline&refresh_mode=paused&storage=flex_tier&stream_sort=time%2Casc&top_n=10&top_o=top&viz=timeseries&x_missing=true&from_ts=1714716000000&to_ts=1714722300000&live=false.

(The approach I suggest will be easier if we can be sure that a single leaseholder will be returned from the trace, as per my above comment.)

ShashwatDash23 replied:
Yes, this makes sense. I'll look into doing this instead!

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -335,9 +368,21 @@ func (p *Prober) Start(ctx context.Context, stopper *stop.Stopper) error {
+						recording := collectAndFinish()
+
+						if traceDumpLimiter.ShouldLog() {
+							log.Health.Infof(ctx, "%v", recording)
```
Arul suggests only logging the trace, if the `kvprober` probe fails, as per below:

> We might also want to consider just dumping the entire trace if a kvprober request fails in the logs

Want to try to implement that?

I think it will be wise to put the trace dumping in a second commit. So first commit just does leaseholder logging. Second commit does trace dumping. The SREs in India are good with `git`, so if not clear how to do this rebase, I am sure they can help. The reason to do two commits it to make easy to revert one of two, in case one has a problem. Also, it just makes for a clean git history, which is something CRL values.

ShashwatDash23 replied:
Yes, I'll look into implementing that and also have this trace dumping as a separate commit.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -335,9 +368,21 @@ func (p *Prober) Start(ctx context.Context, stopper *stop.Stopper) error {
+						recording := collectAndFinish()
+
+						if traceDumpLimiter.ShouldLog() {
+							log.Health.Infof(ctx, "%v", recording)
```
@arulajmani, do you have an opinion on what channel we should dump the trace to? We could introduce a new channel just for these traces.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -335,9 +368,21 @@ func (p *Prober) Start(ctx context.Context, stopper *stop.Stopper) error {
-				probeCtx, sp := tracing.EnsureChildSpan(ctx, p.tracer, opName+" - probe")
-				probe(probeCtx, pl)
-				sp.Finish()
+				func() {
```
I think we should introduce a cluster setting in below file called `kvprober.tracing.enabled` that controls whether `kvprober` takes traces at all. If we backport this, we will set it to `false` for safety. If we just release it in a new major version, we will set it to `true`.

https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvprober/settings.go

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -306,11 +308,42 @@ func (p *Prober) Metrics() Metrics {
+// LogLeaseholderInfo searches through the given tracing recording for log messages
+// indicating updated range information, extracts range IDs and leaseholder node IDs,
+// and logs this information only if the leaseholder for a range has been updated.
+func (p *Prober) LogLeaseholderInfo(ctx context.Context, recording tracingpb.Recording) {
```
We need some tests of this new functionality. I suggest two tests:

1. Unit test of `LogLeaseholderInfo`. Prob the best way is to golden a trace in textual form, then parse the trace to get a `tracingpb.Recording`, then pass that into `LogLeaseholderInfo` and assert that expected leaseholder is parsed out of the trace. This test will run quickly, which is nice, but it could go out of date, if the trace format changes over time, which motivates the second test I have in mind...
2. Update one of the existing integration tests (https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvprober/kvprober_integration_test.go) to assert that the leaseholder logging feature is working on an actual cluster. Here is a test that goldens out the logging done by `sqlprober`. Prob can follow same pattern here, tho details will be different, since different logging library: https://github.com/cockroachlabs/managed-service/pull/16806/files#diff-b7d1606dd13ab761d13bbd2d7785b9139bba43c958703e98d36dfb2bbaef9887R48

Happy to talk in person about the details here. Should be fun!

ShashwatDash23 replied:
Thank you! I will work on these once the actual code is working fine.

joshimhoff reviewed (commented):
Really good job figuring this out so quickly. Your code is quite clean too. Some comments.

ShashwatDash23 commented:
> Really good job figuring this out so quickly. Your code is quite clean too. Some comments.

Thank you so much for the feedback! Will address the comment soon and get back to you!

ShashwatDash23 committed:
kvprober: add leaseholder information for ranges in kvprober logs

Previously, leaseholder information for ranges probed
by kvprober was not being logged.

This commit introduces logging of leaseholder details
extracted from kvprober traces. Currently, getting
leaseholder information is a time consuming process.
This aims to improve the debugging process by providing
readily accessible leaseholder data.

Epic: none
Release note: none

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -306,6 +308,26 @@ func (p *Prober) Metrics() Metrics {
+		leaseRegex := regexp.MustCompile(`\[n(\d+)\]`)
+		leaseholder := leaseRegex.FindStringSubmatch(informationLog)
+		// Return leaseholder node ID if found
+		if len(leaseholder) > 0 {
```
I suggest asserting that `leaseholder` is the exact size you expect here.

ShashwatDash23 replied:
Done!

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -335,8 +357,8 @@ func (p *Prober) Start(ctx context.Context, stopper *stop.Stopper) error {
 
-				probeCtx, sp := tracing.EnsureChildSpan(ctx, p.tracer, opName+" - probe")
-				probe(probeCtx, pl)
+				_, sp := tracing.EnsureChildSpan(ctx, p.tracer, opName+" - probe")
```
What is the reason for ths change?

ShashwatDash23 replied:
My bad. I was trying something else and forgot to change this back.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -306,6 +308,26 @@ func (p *Prober) Metrics() Metrics {
+// indicating leaseholder information, extracts leaseholder node ID,
+// and returns this information. Returns an empty string if leaseholder information
+// is not found.
+func (p *Prober) returnLeaseholderInfo(recording tracingpb.Recording) string {
```
nit: Move this somewhere below `readProbeImpl` or similar. (Generally, public top-level functions go at the top.)

ShashwatDash23 replied:
Done!

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -367,6 +389,12 @@ func (p *Prober) readProbeImpl(ctx context.Context, ops proberOpsI, txns proberT
 	p.metrics.ProbePlanAttempts.Inc(1)
+	var finishAndGetRecording = func() tracingpb.Recording { return nil }
+
+	// trace kv prober requests if tracing is enabled
```
I think we can remove this comment. It doesn't say anything that is not obvious from reading the code.

ShashwatDash23 replied:
Yes, removed now.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -419,7 +449,14 @@ func (p *Prober) readProbeImpl(ctx context.Context, ops proberOpsI, txns proberT
 
 	d := timeutil.Since(start)
-	log.Health.Infof(ctx, "kv.Get(%s), r=%v returned success in %v", step.Key, step.RangeID, d)
+	// Extract leaseholder information from the trace recording if enabled
```
We will want this logic for the write probing also.

ShashwatDash23 replied:
Yes, I have added it for write probes too now. Thank you!

joshimhoff commented on pkg/kv/kvprober/kvprober_integration_test.go:
```diff
@@ -92,6 +94,55 @@ func TestProberDoesReadsAndWrites(t *testing.T) {
 		require.Zero(t, p.Metrics().ProbePlanFailures.Count())
 	})
 
+	t.Run("happy path with tracing enabled", func(t *testing.T) {
```
I think you can just add the new logic to the existing test with case name "happy path".

ShashwatDash23 replied:
Done!

joshimhoff commented on pkg/kv/kvprober/settings.go:
```diff
@@ -134,3 +134,9 @@ var quarantineWriteInterval = settings.RegisterDurationSetting(
+var tracingEnabled = settings.RegisterBoolSetting(
+	settings.ApplicationLevel,
+	"kv.prober.tracing.enabled",
+	"whether the KV prober should collect traces",
```
"whether the KV prober should collect traces" -> "whether the KV prober should collect traces, in order to log info about leaseholders of ranges probed"

ShashwatDash23 replied:
Changed, thank you!

joshimhoff commented on pkg/util/tracing/tracingpb/recording.go:
```diff
@@ -219,6 +219,25 @@ func (r Recording) FindLogMessage(pattern string) (string, bool) {
+//
+// This method strips the redaction markers from all the log messages, which is
+// pretty inefficient.
+func (r Recording) FindLastLogMessage(pattern string) (string, bool) {
```
With only one usage, I think just make this a private function in `kvprober.go` for now. If KV folks think it should be here, they will ask you to move it, but then you'll need more tests, etc.

ShashwatDash23 replied:
I have moved it to `kvprober.go` for now.

joshimhoff commented on pkg/util/tracing/tracingpb/recording.go:
```diff
@@ -219,6 +219,25 @@ func (r Recording) FindLogMessage(pattern string) (string, bool) {
+	for i := len(r) - 1; i >= 0; i-- {
+		sp := r[i]
+		for j := len(sp.Logs) - 1; j >= 0; j-- {
+			msg := sp.Logs[j].Msg().StripMarkers()
```
Can you explain why `StripMarkers` is needed?

ShashwatDash23 replied:
There was a similar function just above looking for logs, so I followed the same pattern. No other particular reason.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -20,6 +20,7 @@ package kvprober
 import (
 	"context"
 	"math/rand"
+	"regexp"
```
Is your plan to merge this PR with no full trace dumping in case of errors & then add that in a follow up PR, if time permits? That seems like a good plan. Just confirming.

ShashwatDash23 replied:
Yes, that's the plan. Ideally should be a small addition.

joshimhoff commented on pkg/kv/kvprober/kvprober_test.go:
```diff
@@ -269,6 +270,78 @@ func TestWriteProbe(t *testing.T) {
 	})
 }
 
+func TestReturnLeaseholderInfo(t *testing.T) {
```
Great test!

ShashwatDash23 replied:
Thank you!

joshimhoff commented on pkg/kv/kvprober/kvprober_integration_test.go:
Great test!

ShashwatDash23 replied:
Thank you!!

joshimhoff commented on pkg/kv/kvprober/kvprober_integration_test.go:
I'm not sure if this read-logs-from-disk approach will be approved by KV. I think we may need an assertion made against an in-memory log buffer. We may not want to depend on reads & writes done from actual disks from within tests. I'm not confident about this tho. We will see what KV says. I will do some research in the meantime about the in-memory-log-buffer approach.

ShashwatDash23 replied:
That's true. I will also have a look in the meantime. Thank you!

joshimhoff replied:
I talked to David from o11y team, and he says the pattern to use in the CRDB repo is to implement an "interceptor". API here: https://github.com/cockroachdb/cockroach/blob/045acd250684d5ca2497478e34a84c4ad1d5699b/pkg/util/log/intercept.go#L26. Example usage here: https://github.com/cockroachdb/cockroach/blob/eccb77d2320955b477bb03bc1f1aee137bd1de76/pkg/ccl/telemetryccl/telemetry_logging_test.go#L158

ShashwatDash23 replied:
Done now!

joshimhoff reviewed (commented):
This looks really good! Great job!!! I'm gonna try to find a reviewer on KV for this now.

ShashwatDash23 commented:
Thank you so much for the review!! I have addressed all the comments.

arulajmani reviewed (commented):
Great work @ShashwatDash23! I've left a few minor comments, but otherwise this looks great.

> Reviewed 3 of 7 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126507#-:-O1rP3e82a3MYMcBiMdW:bb6ox1c)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @joshimhoff and @ShashwatDash23)

___
*[`pkg/kv/kvprober/kvprober.go` line 380 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126507#gh-1662509657:-O1rKtSV8ICk5ElW_SDG:b-kiozbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ccf2138858a971a2e063c64c0955ebdb22952b/pkg/kv/kvprober/kvprober.go#L380)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

@arulajmani, do you have an opinion on what channel we should dump the trace to? We could introduce a new channel just for these traces.
</blockquote></details>

A new channel sounds great. That way, we can use it to dump the entire trace as well without polluting any of the existing logging channels. 
___
*[`pkg/kv/kvprober/kvprober.go` line 540 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126507#-O1rMVp12G3b8ijn3zOL:-O1rMVp2BgmiSWRkGvow:b-p823ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/039e0f3b69b31d2756808ab5dced66bdb06e0a1a/pkg/kv/kvprober/kvprober.go#L540)):*
> ```Go
> // is not found.
> func (p *Prober) returnLeaseholderInfo(recording tracingpb.Recording) string {
> 	// Find the log message containing information about current leaseholder
> ```

nit: missing period, here and other places in this file.
___
*[`pkg/kv/kvprober/kvprober.go` line 541 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126507#-O1rMbHREZO9Tuw6ea0w:-O1rMbHREZO9Tuw6ea0x:bx3cfmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/039e0f3b69b31d2756808ab5dced66bdb06e0a1a/pkg/kv/kvprober/kvprober.go#L541)):*
> ```Go
> func (p *Prober) returnLeaseholderInfo(recording tracingpb.Recording) string {
> 	// Find the log message containing information about current leaseholder
> 	informationLog, isPresent := p.findLastLogMessage(recording, "node received request")
> ```

Let's add a comment explaining this logic. It might help to add a trace snipped (with the non-relevant bits omitted) to explain how we decided upon it.

The logic here seems correct as long as the KV prober request is successfully evaluated by this node. However, if the request isn't evaluated successfully, we can't always make strong claims that the last node to receive the request was the leaseholder. So, in some sense, this is best effort. 

I have a couple of suggestions:
1. Let's follow this up with a full trace. Having that context will be incredibly useful.
2. Let's soften up the language when printing this out to say "likely leaseholder" instead of just "leaseholder" 
___
*[`pkg/kv/kvprober/settings.go` line 143 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126507#-O1rLIQ090e6JuQV-69n:-O1rLIQ090e6JuQV-69o:b-u6vgwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/039e0f3b69b31d2756808ab5dced66bdb06e0a1a/pkg/kv/kvprober/settings.go#L143)):*
> ```Go
> 	"whether the KV prober should collect traces, in order to log info about"+
> 		"leaseholders of ranges probed",
> 	false)
> ```

Did we want to default this to true?
___
*[`pkg/kv/kvprober/kvprober_integration_test.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126507#-O1rLbeoALZaRhwhOmwa:-O1rLbeoALZaRhwhOmwb:b-1ebmsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/039e0f3b69b31d2756808ab5dced66bdb06e0a1a/pkg/kv/kvprober/kvprober_integration_test.go#L71)):*
> ```Go
> 
> 	t.Run("happy path", func(t *testing.T) {
> 		// Define a temporary logging directory for the kv prober logs to be written to
> ```

nit: missing period, here and other comments in this file.
___
*[`pkg/kv/kvprober/kvprober_test.go` line 276 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126507#-O1rLou16_zyV9hEgSaD:-O1rLou16_zyV9hEgSaE:bbhn6jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/039e0f3b69b31d2756808ab5dced66bdb06e0a1a/pkg/kv/kvprober/kvprober_test.go#L276)):*
> ```Go
> 
> 	ctx := context.Background()
> ```

nit: You could consider rewriting this test in a table driven format. See https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Test-Tables. 


<!-- Sent from Reviewable.io -->

ShashwatDash23 commented:
> nit: missing period, here and other places in this file.

Fixed.

--- 
> Let's add a comment explaining this logic. It might help to add a trace snipped (with the non-relevant bits omitted) to explain how we decided upon it.

Added now!

---
> I have a couple of suggestions:
> 1. Let's follow this up with a full trace. Having that context will be incredibly useful.
> 2. Let's soften up the language when printing this out to say "likely leaseholder" instead of just "leaseholder"

Thank you for the comment! It makes alot of sense. Regarding 1. that will come in a separate PR and hopefully I can add it before my internship ends.
2. Changed!

--- 
> Did we want to default this to true?

Based on an earlier comment from Josh, 
> I think we should introduce a cluster setting in below file called kvprober.tracing.enabled that controls whether kvprober takes traces at all. If we backport this, we will set it to false for safety. If we just release it in a new major version, we will set it to true.

and also considering kvprober read and writes are false by default, I felt it was more suitable to keep it false by default. 

--- 
> nit: missing period, here and other comments in this file.

FIxed.

---
> nit: You could consider rewriting this test in a table driven format. See https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Test-Tables.

Thank you for the suggestion! I reviewed the proposed changes but would prefer to skip this refactor at the moment, as I am nearing the end of my internship.

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
"read probe" -> "write probe"

joshimhoff commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -500,6 +532,53 @@ func (p *Prober) quarantineProbe(ctx context.Context, pl planner) {
+	return ""
+}
+
+// FindLastLogMessage returns the last log message in the recording that matches
```
"FindLastLogMessage" -> "findLastLogMessage"

joshimhoff reviewed (approved):
LGTM! Great job. If you have trouble merging, LMK. Happy to help, since close to internship end.

ShashwatDash23 commented:
bors r+

yuzefovich commented:
This fails a recently added linter.

bors r-

ShashwatDash23 commented:
bors r+
